incubator/exporter-prometheus/src/main/java/io/aklivity/zilla/runtime/exporter/prometheus/internal/processor/MetricsProcessor.java

```diff
@@ -0,0 +1,218 @@
+ */
+package io.aklivity.zilla.runtime.exporter.prometheus.internal.processor;
+
+import static io.aklivity.zilla.runtime.engine.metrics.Metric.Kind.COUNTER;
+import static io.aklivity.zilla.runtime.engine.metrics.Metric.Kind.GAUGE;
+import static io.aklivity.zilla.runtime.engine.metrics.Metric.Kind.HISTOGRAM;
+import static io.aklivity.zilla.runtime.exporter.prometheus.internal.layout.HistogramsLayout.BUCKETS;
+import static io.aklivity.zilla.runtime.exporter.prometheus.internal.layout.HistogramsLayout.BUCKET_LIMITS;
+
+import java.io.PrintStream;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Map;
+import java.util.function.Function;
+import java.util.function.LongPredicate;
+import java.util.function.LongSupplier;
+import java.util.stream.Collectors;
+
+import io.aklivity.zilla.runtime.engine.metrics.Metric;
+import io.aklivity.zilla.runtime.exporter.prometheus.internal.labels.LabelManager;
+import io.aklivity.zilla.runtime.exporter.prometheus.internal.layout.MetricsLayout;
+import io.aklivity.zilla.runtime.exporter.prometheus.internal.record.CounterGaugeRecord;
+import io.aklivity.zilla.runtime.exporter.prometheus.internal.record.HistogramRecord;
+import io.aklivity.zilla.runtime.exporter.prometheus.internal.record.MetricRecord;
+
+public class MetricsProcessor
+{
+    private static final long[][] EMPTY = new long[0][0];
+
+    private final Map<Metric.Kind, List<MetricsLayout>> layouts;
+    private final LabelManager labels;
+    private final Function<String, String> kindSupplier;
+    private final Function<String, String> nameSupplier;
+    private final Function<String, String> descriptionSupplier;
+    private final LongPredicate filter;
+    private final List<MetricRecord> metricRecords;
+
+    public MetricsProcessor(
+        Map<Metric.Kind, List<MetricsLayout>> layouts,
+        LabelManager labels,
+        Function<String, String> kindSupplier,
+        Function<String, String> nameSupplier,
+        Function<String, String> descriptionSupplier,
+        String namespaceName,
+        String bindingName)
+    {
+        this.layouts = layouts;
+        this.labels = labels;
+        this.kindSupplier = kindSupplier;
+        this.nameSupplier = nameSupplier;
+        this.descriptionSupplier = descriptionSupplier;
+        this.filter = filterBy(namespaceName, bindingName);
+        this.metricRecords = new LinkedList<>();
+    }
+
+    public void print(
+        PrintStream out)
+    {
+        if (metricRecords.isEmpty())
+        {
+            collectCounters();
+            collectGauges();
+            collectHistograms();
+        }
+        updateRecords();
+        printRecords(out);
+    }
+
+    private LongPredicate filterBy(
+        String namespace,
+        String binding)
+    {
+        int namespaceId = namespace != null ? Math.max(labels.supplyLabelId(namespace), 0) : 0;
+        int bindingId = binding != null ? Math.max(labels.supplyLabelId(binding), 0) : 0;
```

Comment:
Can we add `EngineContext.supplyNamspacedId(String namespace, String local)` instead?

Reply:
I looked into this but it wasn't a trivial fix. The thing is that we don't use this method in the exporter at all, it was just copy-pasted from the metrics command. I decided the remove the implementation for now as it's not used here after all. I suggest that we revisit this issue when we are reconciling the code duplications.